drivers/xen/gnttab.c

```diff
@@ -0,0 +1,336 @@
+{
+	void *page;
+	unsigned long mfn;
+	grant_ref_t gref;
+
+	__ASSERT_NO_MSG(map != NULL);
+
+	page = k_aligned_alloc(XEN_PAGE_SIZE, XEN_PAGE_SIZE);
+	if (page == NULL) {
+		return -ENOMEM;
+	}
+
+	mfn = virt_to_mfn(page);
+	gref = gnttab_grant_access(0, mfn, 0);
+
+	*map = page;
+
+	return gref;
+}
+
+static void gop_eagain_retry(int cmd, struct gnttab_map_grant_ref *gref)
+{
+	unsigned int step = 10, delay = step;
+	int16_t *status = &gref->status;
+
+	do {
+		HYPERVISOR_grant_table_op(cmd, gref, 1);
+		if (*status == GNTST_eagain)
+			k_sleep(K_MSEC(delay));
+
+		delay += step;
+	} while ((*status == GNTST_eagain) && (delay < GOP_RETRY_DELAY));
+
+	if (delay >= GOP_RETRY_DELAY) {
+		LOG_ERR("Failed to map grant, timeout reached\n");
+		*status = GNTST_bad_page;
+	}
+}
+
+int gnttab_map_refs(struct gnttab_map_grant_ref *map_ops, unsigned int count)
+{
+	int i, ret;
+
+	ret = HYPERVISOR_grant_table_op(GNTTABOP_map_grant_ref, map_ops, count);
+	if (ret)
+		return ret;
+
+	for (i = 0; i < count; i++) {
+		switch (map_ops[i].status) {
+		case GNTST_no_device_space:
+			LOG_WRN("map_grant_ref failed, no device space for page #%d\n", i);
+			break;
+
+		case GNTST_eagain:
+			/* Operation not done; need to try again */
+			gop_eagain_retry(GNTTABOP_map_grant_ref, &map_ops[i]);
+			/* Need to re-check status for current page */
+			i--;
+
+			break;
+
+		default:
+			break;
+		}
+	}
+
+	return 0;
+}
+
+int gnttab_unmap_refs(struct gnttab_map_grant_ref *unmap_ops, unsigned int count)
+{
+	return HYPERVISOR_grant_table_op(GNTTABOP_unmap_grant_ref, unmap_ops, count);
+}
+
+
+static const char * const gnttab_error_msgs[] = GNTTABOP_error_msgs;
+
+const char *gnttabop_error(int16_t status)
+{
+	status = -status;
+	if (status < 0 || (uint16_t) status >= ARRAY_SIZE(gnttab_error_msgs)) {
+		return "bad status";
+	} else {
+		return gnttab_error_msgs[status];
+	}
+}
+
+static int gnttab_init(const struct device *d)
+{
+	grant_ref_t gref;
+	struct xen_add_to_physmap xatp;
+	struct gnttab_setup_table setup;
+	xen_pfn_t frames[NR_GRANT_FRAMES];
+	int rc = 0, i;
+
+	/* Will be taken/given during gnt_refs allocation/release */
+	k_sem_init(&gnttab.sem, 0, NR_GRANT_ENTRIES - 1);
+
+	for (
+		gref = GNTTAB_NR_RESERVED_ENTRIES;
+		gref < NR_GRANT_ENTRIES;
+		gref++
+	    ) {
+		put_free_entry(gref);
+	}
+
+	gnttab.table = (grant_entry_v1_t *)
+			DT_REG_ADDR_BY_IDX(DT_INST(0, xen_xen), 0);
+
+	for (i = 0; i < NR_GRANT_FRAMES; i++) {
+		xatp.domid = DOMID_SELF;
+		xatp.size = 0;
+		xatp.space = XENMAPSPACE_grant_table;
+		xatp.idx = i;
+		xatp.gpfn = PFN_DOWN((unsigned long) gnttab.table) + i;
+		rc = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
+		__ASSERT(!rc, "add_to_physmap failed; status = %d\n", rc);
+	}
+
+	setup.dom = DOMID_SELF;
+	setup.nr_frames = NR_GRANT_FRAMES;
+	set_xen_guest_handle(setup.frame_list, frames);
+	rc = HYPERVISOR_grant_table_op(GNTTABOP_setup_table, &setup, 1);
+	__ASSERT(!(rc || setup.status), "Table setup failed; status = %s\n",
```

Comment:
I am not sure, do you think if the `(!rc) && (!setup.status)` or `(rc == 0) && (setup.status == 0)` be more clear and straightforward?

Reply:
Definitely agree with your suggestion, fixed.